velox/adapters/s3fs/S3FileSystem.cpp

```diff
@@ -0,0 +1,267 @@
+  Impl(const Config* config) : config_(config) {}
+
+  std::string getOptionalProperty(
+      const std::string& name,
+      const std::string& defaultValue) {
+    auto value = config_->get(name);
+    if (!value.hasValue()) {
+      return defaultValue;
+    }
+    return value.value();
+  }
+
+  // Configure default AWS credentials provider chain.
+  std::shared_ptr<Aws::Auth::AWSCredentialsProvider>
+  getDefaultCredentialProvider() {
+    return std::make_shared<Aws::Auth::DefaultAWSCredentialsProviderChain>();
+  }
+
+  // Configure with access and secret keys. Used for on-prem.
+  std::shared_ptr<Aws::Auth::AWSCredentialsProvider>
+  getAccessSecretCredentialProvider(
+      std::string accessKey,
+      std::string secretKey) {
+    return std::make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(
+        getAwsString(accessKey), getAwsString(secretKey), getAwsString(""));
+  }
+
+  // Use the input Config parameters and initialize the S3Client
+  void initializeClient() {
+    Aws::Client::ClientConfiguration clientConfig;
+
+    const auto endpoint = getOptionalProperty(S3Config::endpoint, "");
+    clientConfig.endpointOverride = endpoint;
+
+    // default is use SSL
+    const auto useSSL =
+        (getOptionalProperty(S3Config::sslEnabled, "true") == "true");
+    if (useSSL) {
+      clientConfig.scheme = Aws::Http::Scheme::HTTPS;
+    } else {
+      clientConfig.scheme = Aws::Http::Scheme::HTTP;
+    }
+
+    // use virtual addressing for S3 on AWS
+    const auto pathAccessStyle =
+        getOptionalProperty(S3Config::pathAccessStyle, "false");
+    const bool useVirtualAddressing = (pathAccessStyle == "false");
+
+    const auto accessKey = getOptionalProperty(S3Config::accessKey, "");
+    const auto secretKey = getOptionalProperty(S3Config::secretKey, "");
+    const auto useInstanceCred =
+        getOptionalProperty(S3Config::useInstanceCredentials, "");
+    std::shared_ptr<Aws::Auth::AWSCredentialsProvider> credentials_provider;
+    if (accessKey != "" && secretKey != "" && useInstanceCred != "true") {
+      credentials_provider =
+          getAccessSecretCredentialProvider(accessKey, secretKey);
+    } else {
+      credentials_provider = getDefaultCredentialProvider();
+    }
+
+    client_ = std::make_shared<Aws::S3::S3Client>(
+        credentials_provider,
+        clientConfig,
+        Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Never,
+        useVirtualAddressing);
+  }
+
+  std::shared_ptr<Aws::S3::S3Client> getS3Client() {
+    return client_;
+  }
+
+ private:
+  const Config* config_;
+  std::shared_ptr<Aws::S3::S3Client> client_;
+};
+
+S3FileSystem::S3FileSystem(std::shared_ptr<const Config> config)
+    : FileSystem(config) {
+  impl_ = std::make_shared<Impl>(config.get());
+}
+
+void S3FileSystem::initializeClient() {
+  impl_->initializeClient();
+}
+
+std::unique_ptr<ReadFile> S3FileSystem::openFileForRead(std::string_view path) {
+  // Remove the S3://
+  const std::string bucket =
```

Comment:
don't you have a nice helper function for that already in the top of this file?

Reply:
The helper function only checks the scheme. We need the remaining substr here.